mail_outbound_static/models/ir_mail_server.py

```diff
@@ -0,0 +1,146 @@
+        if self.smtp_from:
+            match = re.match(
+                r"^[_a-z0-9-]+(\.[_a-z0-9-]+)*@[a-z0-9-]+(\.[a-z0-9-]+)*(\."
+                r"[a-z]{2,4})$",
+                self.smtp_from,
+            )
+            if match is None:
+                raise ValidationError(_("Not a valid Email From"))
+
+    def _is_valid_domain(self, domain_name):
+        domain_regex = (
+            r"(([\da-zA-Z])([_\w-]{,62})\.){,127}(([\da-zA-Z])"
+            r"[_\w-]{,61})?([\da-zA-Z]\.((xn\-\-[a-zA-Z\d]+)|([a-zA-Z\d]{2,})))"
+        )
+        domain_regex = "{}$".format(domain_regex)
+        valid_domain_name_regex = re.compile(domain_regex, re.IGNORECASE)
+        domain_name = domain_name.lower().strip()
+        return True if re.match(valid_domain_name_regex, domain_name) else False
+
+    @api.model
+    def _get_domain_whitelist(self, domain_whitelist_string):
+        res = domain_whitelist_string.split(",") if domain_whitelist_string else []
+        res = [item.strip() for item in res]
+        return res
+
+    @api.model
+    def send_email(
+        self, message, mail_server_id=None, smtp_server=None, *args, **kwargs
+    ):
+        # Get email_from and name_from
+        if message["From"].count("<") > 1:
+            split_from = message["From"].rsplit(" <", 1)
+            name_from = split_from[0]
+            email_from = split_from[-1].replace(">", "")
+        else:
+            name_from, email_from = parseaddr(message["From"])
+
+        email_domain = email_from.split("@")[1]
+
+        # Replicate logic from core to get mail server
+        # Get proper mail server to use
+        if not smtp_server and not mail_server_id:
+            mail_server_id = self._get_mail_sever(email_domain)
+
+        # If not mail sever defined use smtp_from defined in odoo config
+        if mail_server_id:
+            mail_server = self.sudo().browse(mail_server_id)
+            domain_whitelist = mail_server.domain_whitelist
+            smtp_from = mail_server.smtp_from
+        else:
+            domain_whitelist = tools.config.get("smtp_domain_whitelist")
+            smtp_from = tools.config.get("smtp_from")
+
+        domain_whitelist = self._get_domain_whitelist(domain_whitelist)
+
+        # Replace the From only if needed
+        if smtp_from and (not domain_whitelist or email_domain not in domain_whitelist):
+            email_from = formataddr((name_from, smtp_from))
+            message.replace_header("From", email_from)
+            bounce_alias = (
+                self.env["ir.config_parameter"].sudo().get_param("mail.bounce.alias")
+            )
+            if not bounce_alias:
+                # then, bounce handling is disabled and we want
+                # Return-Path = From
+                if "Return-Path" in message:
+                    message.replace_header("Return-Path", email_from)
+                else:
+                    message.add_header("Return-Path", email_from)
```

Comment:
Update of smtp_from and message values should be done by extending _prepare_email_message(), or the updated values get overriden in there.  I suppose the same issue exists in 15.0.

Reply:
@yostashiro It is best if you forward port #1051 to 16.0